GAWB-2889: lenient library sort + helpful errors on search

davidangb opened the pull request:
no jira, found by Andrew. Two changes in this PR for the library search endpoint:

1. treat `"sortField": ""` in the payload the same as if the user omitted `"sortField"`: the search applies no explicit sort.
2. when ES encounters an exception in its query, make an attempt to bubble the ES error message up to the user. Previously, this endpoint just gave a plain 500 error with no ErrorReport.

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've followed [the instructions](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

davidangb committed:
dont throw exception if user specifies empty string for library sort

davidangb committed:
bubble ES exceptions up to search responses

kyuksel commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala:
```diff
@@ -218,14 +218,17 @@ class LibraryService (protected val argUserInfo: UserInfo,
+          throw new FireCloudExceptionWithErrorReport(ErrorReport(errorMessageFromSearchException(ex)))
+      }
     }
   }
```
In case it's clearer:

```scala
  def findDocuments(criteria: LibrarySearchParams): Future[PerRequestMessage] = {
    val request = for {
      userGroups <- getEffectiveDiscoverGroups(rawlsDAO)
      docs <- searchDAO.findDocuments(criteria, userGroups)
      ids <- rawlsDAO.getWorkspaces
    } yield RequestComplete(updateAccess(docs, ids))

    request recover {
      case ex =>
        throw new FireCloudExceptionWithErrorReport(ErrorReport(errorMessageFromSearchException(ex)))
    }
  }
```

davidangb replied:
The current syntax is intentional (and I've added a code comment to that effect). For-yields are, if I understand correctly, serialized. By declaring the `docsFuture` and `idsFuture` before referencing them in the for-yield, they can be parallelized. The `getEffectiveDiscoverGroups` call *needs* to come first, because its output is used as input to `findDocuments` - so it goes outside the for-yield entirely.

kyuksel replied:
I see. Thanks for the clarification!

kyuksel commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala:
```diff
@@ -239,4 +242,19 @@ class LibraryService (protected val argUserInfo: UserInfo,
   }
+
+  private def errorMessageFromSearchException(ex: Throwable): String = {
+    // elasticsearch errors are often nested, try to dig into them safely to find a message
```
Are the errors at most two-level-deep? I'm not familiar with ElasticSearch but I wonder if there's a generic (and hopefully more idiomatic) way of retrieving the message.

davidangb replied:
I honestly don't know. I tried all the error cases I could think of, like setting a negative number for the search `size` param, or setting a `sortField` that isn't a real field in ES, and in those attempts I saw ES errors nested two-deep. The first level is "all shards failed" and then the second level indicates *how* the shard failed. So it seems better, but I haven't done a comprehensive audit.

kyuksel commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala:
```diff
@@ -239,4 +242,19 @@ class LibraryService (protected val argUserInfo: UserInfo,
+        firstCause.getMessage
+      }
+    } else {
+      ex.getMessage
```
Should we worry about `getMessage`s returning `null` as well?

davidangb replied:
I don't think it will be null - but good call and I've updated the code to be super safe.

kyuksel commented:
There's a typo in `sortField` descriptions of `/api/library/search` and `/api/library/suggest` Swaggers' `Model` sections where `so` should be `to`. It may be good to correct that while at it.

kyuksel commented on src/test/scala/org/broadinstitute/dsde/firecloud/integrationtest/SortSpec.scala:
```diff
@@ -36,7 +36,14 @@ class SortSpec extends FreeSpec with Matchers with BeforeAndAfterAll with LazyLo
+        // results are sorted by relevancy/native index order, which we won't test here
+      }
+    }
+    "search with empty string as sort criteria" - {
```
`"search with empty string as sort criteria"` seems to be a duplicate of `"search with no sort (or filter) criteria"`. Was it meant to input empty strings instead of `None`s?

davidangb replied:
aaah good catch! Fixed.

kyuksel reviewed (commented):
`http://local.broadinstitute.org/#library` on my UI pointing to the changes results in `POST http://localhost:8080/api/library/search 400 (Bad Request)`. This doesn't happen with the `develop` branch so something might have been broken but I didn't get a chance to investigate.

davidangb committed:
typo in swagger doc

davidangb committed:
fix integration test for sorting by empty string

davidangb committed:
super cautious about nulls in error messages

davidangb committed:
code comment

davidangb commented:
@kyuksel back to you for re-review. I've addressed all your comments either with code change or with a reply. I did fix the typo in swagger too. 

Re: the 400 error you were getting - this branch needed a rebase (develop was moving pretty fast!). It's working for me now.

kyuksel reviewed (approved):
Looks great!
